include/blas_quda.h

```diff
@@ -28,261 +28,308 @@ namespace quda {
+       @param[in,out] y update vector set
+       @param[in,out] z update vector set
+       @param[in,out] w update vector set
     */
-    void tripleCGUpdate(double a, double b, const ColorSpinorField &x,
-			ColorSpinorField &y, ColorSpinorField &z, ColorSpinorField &w);
+    void tripleCGUpdate(cvector<double> &a, cvector<double> &b, cvector_ref<const ColorSpinorField> &x,
+                        cvector_ref<ColorSpinorField> &y, cvector_ref<ColorSpinorField> &z,
+                        cvector_ref<ColorSpinorField> &w);
 
     // reduction kernels - defined in reduce_quda.cu
 
     /**
        @brief Compute the maximum absolute real element of a field
        @param[in] a The field we are reducing
     */
-    double max(const ColorSpinorField &x);
+    cvector<double> max(cvector_ref<const ColorSpinorField> &x);
 
     /**
        @brief Compute the maximum real-valued deviation between two
        fields.
        @param[in] x The field we want to compare
        @param[in] y The reference field to which we are comparing against
     */
-    array<double, 2> max_deviation(const ColorSpinorField &x, const ColorSpinorField &y);
+    cvector<array<double, 2>> max_deviation(cvector_ref<const ColorSpinorField> &x,
+                                            cvector_ref<const ColorSpinorField> &y);
+
+    inline array<double, 2> max_deviation(const ColorSpinorField &x, const ColorSpinorField &y)
+    {
+      return max_deviation(cvector_ref<const ColorSpinorField>(x), cvector_ref<const ColorSpinorField>(y));
+    }
 
     /**
        @brief Compute the L1 norm of a field
        @param[in] x The field we are reducing
     */
-    double norm1(const ColorSpinorField &x);
+    cvector<double> norm1(cvector_ref<const ColorSpinorField> &x);
 
     /**
        @brief Compute the L2 norm (||x||^2) of a field
        @param[in] x The field we are reducing
     */
-    double norm2(const ColorSpinorField &x);
+    cvector<double> norm2(cvector_ref<const ColorSpinorField> &x);
+
+    inline double norm2(const ColorSpinorField &x) { return norm2(cvector_ref<const ColorSpinorField> {x}); }
 
     /**
        @brief Compute y += a * x and then (x, y)
        @param[in] a scalar multiplier
        @param[in] x input vector
        @param[in,out] y update vector
     */
-    double axpyReDot(double a, const ColorSpinorField &x, ColorSpinorField &y);
+    cvector<double> axpyReDot(cvector<double> &a, cvector_ref<const ColorSpinorField> &x,
+                              cvector_ref<ColorSpinorField> &y);
 
     /**
        @brief Compute the real-valued inner product (x, y)
        @param[in] x input vector
        @param[in] y input vector
     */
-    double reDotProduct(const ColorSpinorField &x, const ColorSpinorField &y);
+    cvector<double> reDotProduct(cvector_ref<const ColorSpinorField> &x, cvector_ref<const ColorSpinorField> &y);
+
+    inline double reDotProduct(const ColorSpinorField &x, const ColorSpinorField &y)
```

Comment:
What's the reason why we needed an explicit non-`cvector` definition here? Is this something that needed to temporarily go in before a separate outer-product-of-inner-product namespace was added?

Reply:
I added the explicit wrappers to a few functions because there's many cases where the function appears directly in a `printfQuda` statement, and there's no way for the compiler to know to convert it automatically to a scalar, even though we have the cast operator.  Least intrusive thing to do was to just add a few wrappers.

Note that this is actually fixed properly in the reproducible reduction branch that I really should dust off and get in shape for a PR....